### PR TITLE
Feature/13 Issuing Membership VC to Participant Identity Hub

### PR DESCRIPTION
**Pull request by algattik**
## What this PR changes/adds

Implements ADR from https://github.com/agera-edc/RegistrationService/pull/20.

 During participant onboarding, the Registration Service issues a dataspace membership verifiable credential (VC) and store it in the participant's Identity Hub.

 The participant's DID Document must contain a Service Endpoint of type `IdentityHub`, indicating the URL at which the Identity Hub API is available. Access to Identity Hub is currently not authenticated.

 After a participant has been authorised, the Registration Service creates a verifiable credential in JWT format (signed as JWS) and pushes it to the participant's Identity Hub endpoint. After that, the participant is fully onboarded.

## Why it does that

 The membership credential can be used by participants' EDC policy engine to restrict assets to dataspace members.

## Further notes

- System tests must stand up an identity hub to send credentials to. For this purpose, added a launcher to spin a basic EDC with identity hub, and run it in docker-compose
- Retry logic is not implemented yet, so any failure to resolve the DID Document or send VC to Identity Hub will cause the participant enrollment to reach `FAILED` state immediately (participant can submit the request again)

## Linked Issue(s)

https://github.com/agera-edc/MinimumViableDataspace/issues/13

## Checklist

- [ ] added appropriate tests?
- [ ] performed checkstyle check locally?
- [ ] added/updated copyright headers?
- [ ] documented public classes/methods?
- [ ] added/updated relevant documentation?
- [ ] added relevant details to the changelog? (_skip with label `no-changelog`_)
- [ ] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-dataspaceconnector/RegistrationService/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [styleguide](https://github.com/eclipse-dataspaceconnector/RegistrationService/blob/main/styleguide.md) for details_)

### CODE REVIEW

File: extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/credential/VerifiableCredentialService.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ *       Microsoft Corporation - initial API and implementation
+ *
+ */
+
+package org.eclipse.dataspaceconnector.registration.credential;
+
+import org.eclipse.dataspaceconnector.registration.authority.model.Participant;
+import org.eclipse.dataspaceconnector.spi.response.StatusResult;
+
+public interface VerifiableCredentialService {
```
Comment (marcgs) [inline extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/credential/VerifiableCredentialService.java:24]: Missing javadoc

File: extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/credential/VerifiableCredentialService.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    StatusResult<Void> publishVerifiableCredential(Participant participant);
```
Comment (marcgs) [inline extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/credential/VerifiableCredentialService.java]: I think `pushVerifiableCredential` would be a better name as IMO publish implies to make it public somewhere (which it won't be when IH has authentication in place)

File: extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/credential/VerifiableCredentialServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            return StatusResult.failure(FATAL_ERROR, e.toString());
+        }
+        monitor.info("Created dataspace membership VC for " + subject);
+
+        var did = participant.getDid();
+        var didDocument = resolverRegistry.resolve(did);
+        if (didDocument.failed()) {
+            return StatusResult.failure(ERROR_RETRY, "Failed to resolve DID " + did + ". " + didDocument.getFailureDetail());
+        }
+        var hubBaseUrl = getIdentityHubBaseUrl(didDocument.getContent());
```
Comment (marcgs) [inline extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/credential/VerifiableCredentialServiceImpl.java]: I would add more `monitor.debug` statements at each step to indicate what work is being done as this helps a lot when debugging issues. See https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/extensions/iam/decentralized-identity/identity-did-service/src/main/java/org/eclipse/dataspaceconnector/identity/DecentralizedIdentityService.java#L70 for an example.

File: system-tests/src/test/java/org/eclipse/dataspaceconnector/registration/client/RegistrationApiClientTest.java [hunk 4]
```diff
@@ -48,4 +62,27 @@ void listParticipants() {
...
+    }
+
+    private Collection<SignedJWT> getVerifiableCredentialsFromIdentityHub() {
+        var result = identityHubClient.getVerifiableCredentials("http://localhost:8181/api/identity-hub");
+        assertThat(result.succeeded()).isTrue();
+        return result.getContent();
+    }
+
+    private void assertIssuedVerifiableCredential(SignedJWT jwt) throws ParseException {
+        assertThat(jwt.getJWTClaimsSet().getSubject()).isEqualTo(CLIENT_DID_WEB);
```
Comment (marcgs) [inline system-tests/src/test/java/org/eclipse/dataspaceconnector/registration/client/RegistrationApiClientTest.java:85]: Verify "memberOfDataspace" claim?

File: system-tests/README.md [hunk 5]
```diff
@@ -4,10 +4,10 @@ These tests deploy the application in a docker environment, and use the generate
 
 ### Running test locally
 
-Build the application launcher:
+Build the application launchers:
 
 ```
-./gradlew :launcher:shadowJar
+./gradlew shadowJar
```
Comment (marcgs) [inline system-tests/README.md]: I get `registration-service    | org.eclipse.dataspaceconnector.core.security.azure.AzureVaultException: Either 'edc.vault.certificate' or 'edc.vault.clientsecret' must be supplied but both were null` when running integration tests locally

Comment (marcgs) [reply system-tests/README.md]: Do you need `./gradlew -DuseFsVault="true" :launcher:shadowJar :system-tests:launchers:identity-hub:shadowJar` here?

File: system-tests/src/test/java/org/eclipse/dataspaceconnector/registration/client/RegistrationApiClientTest.java (hunk 4 above)
Comment (algattik) [reply system-tests/src/test/java/org/eclipse/dataspaceconnector/registration/client/RegistrationApiClientTest.java:85]: Intentionally not done - here we do minimal assertions to ensure a VC was pushed (the assertions are to ensure it is not some other VC). The claim is covered in UT.

**Review by marcgs — APPROVED** [review]

**Review by Izzzu — APPROVED** [review]

**Pull request merged**